File: src/cli/commands/archive.ts
import { defineCommand } from "citty";
import {
  readFileSync,
  readdirSync,
  existsSync,
  mkdirSync,
  renameSync,
  writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { parse as parseYaml } from "yaml";
import { consola } from "consola";
import { resolveConfig } from "../../config/index.js";
import type { IssueYaml } from "../../schemas/issue.js";

export const archiveCommand = defineCommand({
  meta: {
    name: "archive",
    description: "Archive a completed plan",
  },
  args: {
    slug: {
      type: "positional",
      description: "Plan slug to archive",
      required: true,
    },
    force: {
      type: "boolean",
      description: "Skip confirmation and active-issue check",
      default: false,
    },
  },
  async run({ args }) {
    const config = resolveConfig(process.cwd());
    const planDir = join(config.plans_dir, args.slug);

    if (!existsSync(planDir)) {
      console.error(`Plan not found: ${planDir}`);
      process.exit(1);
    }

    const issues = loadIssues(join(planDir, "issues"));

    // Check for active issues
    const active = issues.filter((i) =>
      ["in-progress", "blocked"].includes(i.status),
    );
    if (active.length > 0 && !args.force) {
      console.error(
        `Cannot archive: ${active.length} active issues (${active.map((i) => `#${i.id}`).join(", ")})`,
      );
      console.error("Use --force to override.");
      process.exit(1);
    }

    // Confirm
    if (!args.force) {
      const confirm = await consola.prompt(
        `Archive plan "${args.slug}"?`,
        { type: "confirm" },
      );
      if (!confirm) {
        console.log("Aborted.");
        return;
      }
    }

    // Generate summary
    const doneCount = issues.filter((i) => i.status === "done").length;
    const totalPoints = issues.reduce((s, i) => s + i.points, 0);
    const donePoints = issues
      .filter((i) => i.status === "done")
      .reduce((s, i) => s + i.points, 0);

    const byStatus: Record<string, number> = {};
    for (const issue of issues) {
      byStatus[issue.status] = (byStatus[issue.status] ?? 0) + 1;
    }

    const summary = [
      `# Archive Summary: ${args.slug}`,
      "",
      `Archived: ${new Date().toISOString().slice(0, 10)}`,
      "",
      `## Stats`,
      `- Issues: ${doneCount}/${issues.length} done`,
      `- Points: ${donePoints}/${totalPoints}`,
      "",
      "## By Status",
      ...Object.entries(byStatus).map(([s, c]) => `- ${s}: ${c}`),
    ].join("\n");

    // Move to archive
    const archiveDir = join(config.plans_dir, ".archive");
    const archiveDest = join(archiveDir, args.slug);
    mkdirSync(archiveDir, { recursive: true });
    renameSync(planDir, archiveDest);

    // Write summary
    writeFileSync(join(archiveDest, "summary.md"), summary, "utf-8");

    console.log(`Archived: ${archiveDest}`);
  },
});

function loadIssues(dir: string): IssueYaml[] {
  try {
    return readdirSync(dir)
      .filter((f: string) => f.endsWith(".yaml"))
      .map((f: string) =>
        parseYaml(readFileSync(join(dir, f), "utf-8")) as IssueYaml,
      );
  } catch {
    return [];
  }
}

File: internal/cli/archive.go
package cli

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/jasonraimondi/plan-bender/internal/config"
	"github.com/jasonraimondi/plan-bender/internal/planrepo"
	"github.com/jasonraimondi/plan-bender/internal/schema"
	"github.com/spf13/cobra"
	"gopkg.in/yaml.v3"
)

// NewArchiveCmd creates the archive command.
func NewArchiveCmd() *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:   "archive <slug>",
		Short: "Archive a completed plan",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			slug := args[0]
			root, _ := os.Getwd()

			cfg, err := config.Load(root)
			if err != nil {
				return err
			}

			planDir := filepath.Join(cfg.PlansDir, slug)

			// Hold the planrepo session lock through the entire archive
			// operation so a concurrent dispatcher mid-Commit cannot interleave
			// writes into the directory we are about to rename. Safe because
			// the lock file lives at <plansDir>/.pb-lock — sibling to
			// planDir, not inside it — so the rename does not move the lock
			// file itself, and the held fd stays valid until Close.
			sess, err := planrepo.NewProd(cfg.PlansDir).Open(slug)
			if err != nil {
				return err
			}
			defer func() { _ = sess.Close() }()

			snap, err := sess.Snapshot()
			if err != nil {
				return err
			}
			issues := snap.Issues

			// Check for active issues
			if !force {
				var active []string
				for _, iss := range issues {
					if iss.Status == "in-progress" || iss.Status == "blocked" {
						active = append(active, fmt.Sprintf("#%d (%s)", iss.ID, iss.Status))
					}
				}
				if len(active) > 0 {
					return fmt.Errorf("active issues: %s — use --force to override", strings.Join(active, ", "))
				}
			}

			// Ensure the archive parent dir exists before writing summary so a
			// rename failure isn't preceded by an unrelated mkdir failure.
			archiveDir := filepath.Join(cfg.PlansDir, ".archive")
			if err := os.MkdirAll(archiveDir, 0o755); err != nil {
				return err
			}

			// Write summary atomically so a partial summary.md never lands on
			// disk if the process is killed mid-write.
			summary := buildSummary(slug, issues)
			summaryPath := filepath.Join(planDir, "summary.md")
			if err := planrepo.AtomicWrite(summaryPath, []byte(summary), 0o644); err != nil {
				return fmt.Errorf("writing summary: %w", err)
			}

			// Move to archive. If the rename fails, the freshly-written
			// summary.md is now orphaned in planDir — remove it so a retry
			// doesn't see stale summary content from a prior failed attempt.
			dst := filepath.Join(archiveDir, slug)
			if renameErr := os.Rename(planDir, dst); renameErr != nil {
				cleanupErr := os.Remove(summaryPath)
				return errors.Join(fmt.Errorf("moving to archive: %w", renameErr), cleanupErr)
			}

			if isAgentMode(cmd) {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(map[string]string{
					"status": "ok",
					"file":   dst,
				})
			}
			fmt.Fprintf(cmd.OutOrStdout(), "archived %s to %s\n", slug, dst)
			return nil
		},
	}

	cmd.Flags().BoolVar(&force, "force", false, "archive even with active issues")
	return cmd
}

func buildSummary(slug string, issues []schema.IssueYaml) string {
	var b strings.Builder

	byStatus := make(map[string]int)
	totalPoints := 0
	donePoints := 0
	for _, iss := range issues {
		byStatus[iss.Status]++
		totalPoints += iss.Points
		if iss.Status == "done" {
			donePoints += iss.Points
		}
	}

	b.WriteString(fmt.Sprintf("# Archive: %s\n\n", slug))
	b.WriteString(fmt.Sprintf("Archived: %s\n\n", time.Now().Format("2006-01-02")))
	b.WriteString(fmt.Sprintf("Issues: %d done / %d total\n", byStatus["done"], len(issues)))
	b.WriteString(fmt.Sprintf("Points: %d / %d\n\n", donePoints, totalPoints))
	b.WriteString("## By Status\n\n")

	data, _ := yaml.Marshal(byStatus)
	b.Write(data)

	return b.String()
}
